### DISCUSSION TITLE
Limit switches

### DISCUSSION THREAD

**Discussion by Festivejelly** [kachurovskiy/nanoels#197] [Ideas]
This may have been talked about before but after using the lathe controller for a while I think a limit switch on the X could be super useful.

I normally either take a light cut or I mount one of those LED beeper probe things into the chuck and touch off on that. Both ways get the cutter super close, an sometimes i might move it an extra 0,05mm in or out just to fine tune.

It would be great once getting the Zero to then home on the X and when it hits the microswitch record the value of where the controller thinks it is.

This way we can then put that value in our config and then every time we start up the controller we can simply home the axis, without the need to do test cuts or use probes.

Ive tested the repeatability on some omron switches and they are will within spec to be useful enough.

I dont see this being valuable on the Z axis though since the length of work and the zero is going to be changing all the time.

but on the X axis the zero will always be the centre line.

**Comment by Festivejelly** [comment]
I think maybe one of the AUX inputs could be used? Then like A11 and GND.
Im happy to take a crack at this but I dont want to waste my time if its not going to work :)

**Comment by kachurovskiy** [comment]
Hi @Festivejelly, you can use A11 for digital input - it's pulled up by default and you just need the switch to control the connection to GND as you said.

Perhaps the zeroing cycle could start if you hold the X button and press the Play button.

The fundamental problem with this approach is that your cutting diameter depends on stock stick out, depth of cut, tool post stiffness and screw wear in the particular place where you're doing the cutting. None of those would be taken into account by this microswitch zeroing logic - so precision will suffer and you'd still have to zero from the light cut in the end - rendering the entire change less useful.

It would definitely be interesting for everyone to see how this plays out in reality if you decide to take a stab at it, happy to be proven wrong! Cheers.

**Reply by Festivejelly** [reply]
Agree there are a few parameters that could affect this for sure. But we could have a "profile" that could be applied to compensate for this.

Similar to how ive implemented tool changes where each tool has an offset from the primary tool, an offset could be applied and calibrated. I'll have a crack and see what I get. Much smaller less rigid CNC lathes seem to function well with endstops. Running with a solid tool post would help with this I think. Im actaully really impressed with the repeatability I can get already tbh when I manually calibrate it based on one of these: https://www.amazon.co.uk/gp/product/B07PDZF6DP/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1 in the chuck. But it does mean removing the work and if you use a 4 jaw chuck its a pain.

I'll report back when I have something I can demo.

**Comment by Clive66** [comment]
Following on from this.
 I had a near miss when I accidentally de-activated a soft limit. Would it be possible to wire in a back up mechanical limit switch on the Z axis?  So that in effect if the soft limit fails the limit switch cuts the motor or discables the axis.
Could this be done without major code changes.
Thanks

**Reply by kachurovskiy** [reply]
You could cut the power to the motors or the entire system when such safety switch is hit, doesn't have to be part of the ELS logic?

**Reply by Clive66** [reply]
That makes it a super easy modification. Thanks for the rapid reply!

**Comment by demsarmatjaz** [comment]
I suggest adding a "normally closed" mechanical limit switch between the "ENA" output for the Z-axis on the ELS and the negative pin of the "ENA" input on the stepper motor controller for the Z-axis. The positive pin is already connected to +5V from the ELS, leave that as is. The operating logic is inverted, which is correct in this case. I had to add a pulldown resistor between negative pin of stepper driver ENA input and logical GND - GND from ELS. Start with 100k, and if it's not enough, move towards 10k. I don't recommend anything lower due to the load on the ESP32 output. Mount this switch on a fixture that can be moved left-right to acquire the desired position. I will even use two of them, one on each side of the lathe support. It's better to stop the stepper than to break something. I am using a NEMA 34 4.5Nm stepper for the Z-axis and it's a beast. The same solution can also be used for the X-axis at the most in and out positions to avoid unnecessary wear of the motor and mechanical components.

**Comment by Clive66** [comment]
Thanks for your idea.
![pic switch](https://github.com/user-attachments/assets/24cf6085-915e-4a8a-aa75-87066c269905)
Bracket for limit switch already made and secured with strong magnets. I am more mechanical than electrical. Can you explain breifly the function of a pulldown resistor and how you would arrive at the correct value please. Would it be to prevent voltage spikes?
Advice much appreciated..

**Reply by Clive66** [reply]
Chat GPT provided the answer. Thanks for the input.

**Reply by demsarmatjaz** [reply]
![Limit and pulldown](https://github.com/user-attachments/assets/594d7023-6255-42ff-9697-585f63efe723)
In my case, the pulldown resistor is needed to activate "ENA" input of the driver and so to disable the stepper motor outputs. Somehow confusing, but that's how it is. Imagine that the motor output is only active if the "ENA" input of the driver is not active and vice versa. Input "ENA" (it would be better name it "DISABLE") input is normally activated by setting the ELS "ENA" output to "LOW" but if you cut this signal line the transmitter led diode in the optocoupler of driver "ENA" input will not lit, so we need to obtain a GND through a pulldown resistor connected to the same GND as ELS is connected to. Because ELS sends logical HIGH to "ENA" output for enabling the motor, this resistor must be as high as possible to pull as little as possible current from the SN74HCT245N - steppers are driven through this chip which in this case make level shifting between 3V3 and 5V. This chip can give up to 25mA per pin but, if I am not wrong,  only 6mA if we want to keep output voltage in the range of specified. Theoretically, you could use a resistor down to 1k (5V / 0.006A = 833Ohm, 1kOhm with a small margin and taking into account the consumption of the optocoupler), but 10k will cause less stress to the output of 74245 and still obtain enough current for firing opto. Pulldown could also affect to minimize induced interference on the cable but, anyway, especially if you are using VFD for driving main motor, you must use quality shilded cables and connect those shields to the grounded chassis of the machine all in one point. All cable shields must only be grounded at one side only And don't connect logic GND to chassis GND!

**Reply by Clive66** [reply]
That makes perfect sense thank you. Have ordered a pack of resistors from Amazon!
